src/nvim/window.c

```diff
@@ -1027,9 +1027,11 @@ void ui_ext_win_position(win_T *wp, bool validate)
   }
 }
 
-void ui_ext_win_viewport(win_T *wp)
+void ui_ext_win_viewport(win_T *wp, bool allow_updated)
```

Comment:
Is this argument really needed? If the window needs another redraw after calling `win_update()`, shouldn't `win_viewport` still be delayed?
```suggestion
void ui_ext_win_viewport(win_T *wp)
```

Reply:
It is unfortunately required. `win_viewport` has the dual duty of indicating the row and column, and that is allowed to happen without calling `redraw_later` In those cases `v_viewport_invalid` is set and `ui_ext_win_viewport` is called directly from `UIFlush` and can happen basically everywhere.

So, if the check `allow_updated` is not there, we are back to square one, `win_viewport` gets sent with a flush, while the actual screen is still dirty.

My first approach was actually to remove `win_viewport` from `UIFlush`, but then there will be no feedback when only the `curcolumn` or `curow` are updated. The next approach was to call `win_view_port` unconditionally after `win_update`, but some tests are still failing with that, and it also sounds extremely risky, since it's been called from so many places before, even outside the `update_screen`

Reply:
> `win_viewport` has the dual duty of indicating the row and column, and that is allowed to happen without calling `redraw_later` In those cases `v_viewport_invalid` is set and `ui_ext_win_viewport` is called directly from `UIFlush` and can happen basically everywhere.

This would be useful to have as a `@param` docstring.